pkg/ccl/testccl/sqlccl/sqlccl_test: TestReadCommittedStmtRetry failed

pkg/ccl/testccl/sqlccl/sqlccl_test.TestReadCommittedStmtRetry [failed](https://tanzanite.cluster.engflow.com/invocations/default/520198f4-b95b-4ae9-beb0-ddf92728660b?testReportRun=2&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on release-24.1.0-rc @ [34bbed697e77466f52012a80f0f6ae496c0d2889](https://github.com/cockroachdb/cockroach/commits/34bbed697e77466f52012a80f0f6ae496c0d2889):

Fatal error:

```
panic: test timed out after 41m37s
running tests:
	TestReadCommittedStmtRetry (41m37s)

```
Stack: 

```
goroutine 84012 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x265
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x45
```
<details><summary>Log preceding fatal error</summary>
<p>

```
github.com/cockroachdb/cockroach/pkg/kv/kvpb._Internal_Batch_Handler.func1({0xdbe8dd8, 0xc003e179b0}, {0x9e49540, 0xc006610000})
	github.com/cockroachdb/cockroach/pkg/kv/kvpb/bazel-out/k8-fastbuild/bin/pkg/kv/kvpb/kvpb_go_proto_/github.com/cockroachdb/cockroach/pkg/kv/kvpb/api.pb.go:10632 +0xbf
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.ServerInterceptor.func12({0xdbe8dd8, 0xc003e179b0}, {0x9e49540, 0xc006610000}, 0xc002eea3a0, 0xc00194a7f8)
	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97 +0x72c
google.golang.org/grpc.getChainUnaryHandler.func1({0xdbe8dd8, 0xc003e179b0}, {0x9e49540, 0xc006610000})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1163 +0x12d
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3({0xdbe8dd8, 0xc003e179b0}, {0x9e49540, 0xc006610000}, 0xc002eea3a0, 0xc00723d6c0)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169 +0xf1
google.golang.org/grpc.getChainUnaryHandler.func1({0xdbe8dd8, 0xc003e179b0}, {0x9e49540, 0xc006610000})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1163 +0x12d
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor({0xc002308000?, {{0x0?}, {0xdc34180?, 0xc001179500?}}}, {0xdbe8dd8, 0xc003e178c0}, {0x9e49540, 0xc006610000}, 0xc002eea3a0, 0xc00723d680)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:95 +0x2b7
google.golang.org/grpc.getChainUnaryHandler.func1({0xdbe8dd8, 0xc003e178c0}, {0x9e49540, 0xc006610000})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1163 +0x12d
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1({0xdbe8dd8, 0xc003e178c0})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:136 +0x6b
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc00138dea0, {0xdbe8dd8, 0xc003e178c0}, {0xc00141c808?, 0xc00723d640?}, 0xc0094f1700)
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:337 +0x186
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1({0xdbe8dd8, 0xc003e178c0}, {0x9e49540, 0xc006610000}, 0xc002eea3a0, 0xc00723d640)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:134 +0x151
google.golang.org/grpc.NewServer.chainUnaryServerInterceptors.chainUnaryInterceptors.func1({0xdbe8dd8, 0xc003e178c0}, {0x9e49540, 0xc006610000}, 0xc002eea3a0, 0xc00194a7f8)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xe8
github.com/cockroachdb/cockroach/pkg/kv/kvpb._Internal_Batch_Handler({0x9e18b20, 0xc004727008}, {0xdbe8dd8, 0xc003e178c0}, 0xc009a4e9a0, 0xc001b05aa0)
	github.com/cockroachdb/cockroach/pkg/kv/kvpb/bazel-out/k8-fastbuild/bin/pkg/kv/kvpb/kvpb_go_proto_/github.com/cockroachdb/cockroach/pkg/kv/kvpb/api.pb.go:10634 +0x1f4
google.golang.org/grpc.(*Server).processUnaryRPC(0xc003d55a40, {0xdc3f320, 0xc004fda1a0}, 0xc0077170e0, 0xc001875860, 0x11be5860, 0x0)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1336 +0x1814
google.golang.org/grpc.(*Server).handleStream(0xc003d55a40, {0xdc3f320, 0xc004fda1a0}, 0xc0077170e0, 0x0)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1704 +0x1013
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:965 +0xe5
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 2094
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:963 +0x4a8

goroutine 79143 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc007b640a8, 0x367)
	GOROOT/src/runtime/sema.go:569 +0x159
sync.(*Cond).Wait(0xc007b64098)
	GOROOT/src/sync/cond.go:70 +0x75
github.com/cockroachdb/pebble/record.(*flusherCond).Wait(...)
	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:369
github.com/cockroachdb/pebble/record.(*LogWriter).flushLoop(0xc007b64000, {0xc003eb6cc8?, 0x94efce0?})
	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:643 +0xb79
runtime/pprof.Do({0xdbe8c18, 0x132807c0}, {{0xc000de9e60?, 0x1a1cd00?, 0xc001e74d10?}}, 0xc005321fc0)
	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
github.com/cockroachdb/pebble/record.NewLogWriter.func2()
	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:562 +0x92
created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 133
	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:561 +0x8e5

goroutine 26607 [select]:
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReadCommittedStmtRetry.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38457

--- a/pkg/ccl/testccl/sqlccl/read_committed_test.go
+++ b/pkg/ccl/testccl/sqlccl/read_committed_test.go
@@ -40,16 +40,18 @@ func TestReadCommittedStmtRetry(t *testing.T) {
 	ctx := context.Background()
 	params := base.TestServerArgs{}
 
-	var readCommittedWriteCount atomic.Int64
-	var finishedReadCommittedScans sync.WaitGroup
-	finishedReadCommittedScans.Add(1)
-	var finishedExternalTxn sync.WaitGroup
-	finishedExternalTxn.Add(1)
+	var trapReadCommittedWrites atomic.Bool
+	var trappedReadCommittedWritesOnce sync.Once
+	finishedReadCommittedScans := make(chan struct{})
+	finishedExternalTxn := make(chan struct{})
 	var sawWriteTooOldError atomic.Bool
 	var codec keys.SQLCodec
 	var kvTableId uint32
 
 	filterFunc := func(ctx context.Context, ba *kvpb.BatchRequest) *kvpb.Error {
+		if !trapReadCommittedWrites.Load() {
+			return nil
+		}
 		if ba.Txn == nil || ba.Txn.IsoLevel != isolation.ReadCommitted {
 			return nil
 		}
@@ -61,13 +63,10 @@ func TestReadCommittedStmtRetry(t *testing.T) {
 				if err != nil || tableID != kvTableId {
 					return nil
 				}
-				// Because of the queries the test executes below, we know that before
-				// the second read committed write begins, the read committed scans
-				// will have finished.
-				if newCount := readCommittedWriteCount.Add(1); newCount == 2 {
-					finishedReadCommittedScans.Done()
-					finishedExternalTxn.Wait()
-				}
+				trappedReadCommittedWritesOnce.Do(func() {
+					close(finishedReadCommittedScans)
+					<-finishedExternalTxn
+				})
 			}
 		}
 
@@ -110,6 +109,9 @@ func TestReadCommittedStmtRetry(t *testing.T) {
 	_, err = tx.Exec(`UPDATE kv SET v = v+10 WHERE k = 'a'`)
 	require.NoError(t, err)
 
+	// Start blocking writes in the read committed transaction.
+	trapReadCommittedWrites.Store(true)
+
 	// Perform a series of reads and writes in the second statement.
 	// Read from "b" and "c" to establish refresh spans.
 	// Write to "b" in the transaction, without issue.
@@ -121,7 +123,7 @@ func TestReadCommittedStmtRetry(t *testing.T) {
 	})
 
 	// Wait for the table to be scanned first.
-	finishedReadCommittedScans.Wait()
+	<-finishedReadCommittedScans
 
 	// Write to "c" outside the transaction to create a write-write conflict.
 	_, err = sqlDB.Exec(`UPDATE kv SET v = v+10 WHERE k = 'c'`)
@@ -129,7 +131,7 @@ func TestReadCommittedStmtRetry(t *testing.T) {
 
 	// Now let the READ COMMITTED write go through. It should encounter a
 	// WriteTooOldError and retry.
-	finishedExternalTxn.Done()
+	close(finishedExternalTxn)
 
 	err = g.Wait()
 	require.NoError(t, err)

